File: rsa.js
// Depends on jsbn.js and rng.js

// Version 1.1: support utf-8 encoding in pkcs1pad2

// convert a (hex) string to a bignum object
function parseBigInt(str,r) {
  return new BigInteger(str,r);
}

function linebrk(s,n) {
  var ret = "";
  var i = 0;
  while(i + n < s.length) {
    ret += s.substring(i,i+n) + "\n";
    i += n;
  }
  return ret + s.substring(i,s.length);
}

function byte2Hex(b) {
  if(b < 0x10)
    return "0" + b.toString(16);
  else
    return b.toString(16);
}

// PKCS#1 (type 2, random) pad input string s to n bytes, and return a bigint
function pkcs1pad2(s,n) {
  if(n < s.length + 11) { // TODO: fix for utf-8
    alert("Message too long for RSA");
    return null;
  }
  var ba = new Array();
  var i = s.length - 1;
  while(i >= 0 && n > 0) {
    var c = s.charCodeAt(i--);
    if(c < 128) { // encode using utf-8
      ba[--n] = c;
    }
    else if((c > 127) && (c < 2048)) {
      ba[--n] = (c & 63) | 128;
      ba[--n] = (c >> 6) | 192;
    }
    else {
      ba[--n] = (c & 63) | 128;
      ba[--n] = ((c >> 6) & 63) | 128;
      ba[--n] = (c >> 12) | 224;
    }
  }
  ba[--n] = 0;
  var rng = new SecureRandom();
  var x = new Array();
  while(n > 2) { // random non-zero pad
    x[0] = 0;
    while(x[0] == 0) rng.nextBytes(x);
    ba[--n] = x[0];
  }
  ba[--n] = 2;
  ba[--n] = 0;
  return new BigInteger(ba);
}

// "empty" RSA key constructor
function RSAKey() {
  this.n = null;
  this.e = 0;
  this.d = null;
  this.p = null;
  this.q = null;
  this.dmp1 = null;
  this.dmq1 = null;
  this.coeff = null;
}

function lala(text){
	return "lala-lolo ja " +text;
}

// Set the public key fields N and e from hex strings
function RSASetPublic(N,E) {
  if(N != null && E != null && N.length > 0 && E.length > 0) {
    this.n = parseBigInt(N,16);
    this.e = parseInt(E,16);
  }
  else
    alert("Invalid RSA public key");
}

// Perform raw public operation on "x": return x^e (mod n)
function RSADoPublic(x) {
  return x.modPowInt(this.e, this.n);
}

// Return the PKCS#1 RSA encryption of "text" as an even-length hex string
function RSAEncrypt(text) {
  var m = pkcs1pad2(text,(this.n.bitLength()+7)>>3);
  if(m == null) return null;
  var c = this.doPublic(m);
  if(c == null) return null;
  var h = c.toString(16);
  if((h.length & 1) == 0) return h; else return "0" + h;
}

// Return the PKCS#1 RSA encryption of "text" as a Base64-encoded string
function RSAEncryptB64(text) {
  var h = this.encrypt(text);
  if(h) return hex2b64(h); else return null;
}

// protected
RSAKey.prototype.doPublic = RSADoPublic;

// public
RSAKey.prototype.setPublic = RSASetPublic;
RSAKey.prototype.encrypt = RSAEncrypt;
RSAKey.prototype.encrypt_b64 = RSAEncryptB64;


//my encrypt function, using fixed mudulus
/*var modulus = "BEB90F8AF5D8A7C7DA8CA74AC43E1EE8A48E6860C0D46A5D690BEA082E3A74E1"
			 +"571F2C58E94EE339862A49A811A31BB4A48F41B3BCDFD054C3443BB610B5418B"
			 +"3CBAFAE7936E1BE2AFD2E0DF865A6E59C2B8DF1E8D5702567D0A9650CB07A43D"
			 +"E39020969DF0997FCA587D9A8AE4627CF18477EC06765DF3AA8FB459DD4C9AF3";*/
var modulus = "00c4ad6f19f5588f18e7ce5b5a95347cc1ecdac3b78bf59a53db8a988213333b42c7fb42d4dca03ae061c446740e709da961bc4f99cad2acc8e6c20b10e76349275401d0470fd9c26465554a492881e04ddc9aef4b50f8abf4f9249ff241367c7284bd8891b2c283990e39f816bedab80da4e40280d17d3ba20342a470c4321265";
var publicExponent = "10001";
function MyRSAEncryptB64(text)
{
	var rsa = new RSAKey();
    rsa.setPublic(modulus, publicExponent);
    return rsa.encrypt_b64(text);
}

File: Sms/src/util/RSA.java
/**
 * 
 */
package util;

import java.math.BigInteger;
import java.nio.ByteBuffer;
import java.nio.IntBuffer;
import java.security.SecureRandom;

import org.bouncycastle.util.encoders.Base64;

/**
 * @author Triin
 *
 */
public class RSA {
	BigInteger n;
	BigInteger e;
	// Set the public key fields N and e from hex strings
	public void RSASetPublic(String N,String E) {
	  if(N != null && E != null && N.length() > 0 && E.length() > 0) {
	    this.n = new BigInteger(N,16);
	    this.e = new BigInteger(E,16);
	  }
	  else
	    System.out.println("Invalid RSA public key");
	}
	
	// Return the PKCS#1 RSA encryption of "text" as an even-length hex string
	public String RSAEncrypt(String text) {
	  BigInteger m = pkcs1pad2(text,(this.n.bitLength()+7)>>3);
	  if(m == null) return null;
	  BigInteger c = m.modPow(e, n);
	  if(c == null) return null;
	  String h = new String(Base64.encode(c.toByteArray()));
	  System.out.println("h on: " + h);
	//String h = ConvertHexStringToBase64(new String(bytes));
	  if((h.length() & 1) == 0) return h; else return "0" + h;
	}
	// PKCS#1 (type 2, random) pad input string s to n bytes, and return a bigint
	private BigInteger pkcs1pad2(String s,int n) {
	  if(n < s.length() + 11) { // TODO: fix for utf-8
	    System.out.println("Message too long for RSA");
	    return null;
	  }
	  int[] ba = new int[1000000];
	  int i = s.length() - 1;
	  while(i >= 0 && n > 0) {
	    //var c = s.charCodeAt(i--);
	    int c = Character.codePointAt(s, i--);
	    if(c < 128) { // encode using utf-8
	      ba[--n] = c;
	    }
	    else if((c > 127) && (c < 2048)) {
	      ba[--n] = (c & 63) | 128;
	      ba[--n] = (c >> 6) | 192;
	    }
	    else {
	      ba[--n] = (c & 63) | 128;
	      ba[--n] = ((c >> 6) & 63) | 128;
	      ba[--n] = (c >> 12) | 224;
	    }
	  }
	  ba[--n] = 0;
	  SecureRandom rng = new SecureRandom();
	  byte[] x = new byte[1000000]; //HARDCODED!!!
	  while(n > 2) { // random non-zero pad
	    x[0] = 0;
	    while(x[0] == 0) rng.nextBytes(x);
	    ba[--n] = x[0];
	  }
	  ba[--n] = 2;
	  ba[--n] = 0;
	  //System.out.println("BigInteger.toString: " + ba.toString());
	  return toBigInteger(ba);
	}
	public BigInteger toBigInteger(int[] data) {
	    byte[] array = new byte[data.length * 4];
	    ByteBuffer bbuf = ByteBuffer.wrap(array);
	    IntBuffer ibuf = bbuf.asIntBuffer();
	    ibuf.put(data);
	    return new BigInteger(array);
	}
	// Return the PKCS#1 RSA encryption of "text" as a Base64-encoded string
	
	private static String ConvertHexStringToBase64(String hexString) {
	    System.out.println(hexString);
	    if ((hexString.length()) % 2 > 0)
	          throw new NumberFormatException("Input string was not in a correct format.");
	    byte[] buffer = new byte[hexString.length() / 2];
        int i = 2;
        while (i < hexString.length())
        	{
            buffer[i / 2] = (byte)Integer.parseInt(hexString.substring(i, i + 2),16);
            i += 2;
        }
	    System.out.println("hexSring"+hexString+"afterconverttobase64"+new String(Base64.encode(buffer)));
	    return new String(Base64.encode(buffer));

	}
	
}
